### PR TITLE
Refactor CLI exception handling to use specific httpx exceptions

### LINKED ISSUE #417: Refactor CLI exception handling to use specific httpx exceptions
Follow-up from PR #415 review.

Currently, several CLI commands use broad `except Exception` blocks when making HTTP calls to the server. This pattern catches both expected errors (connection failures, timeouts, HTTP errors) and unexpected programming errors (attribute errors, bugs in parsing), which makes debugging harder and can hide bugs.

## Scope

Refactor CLI commands to catch specific `httpx` exception types:
- `httpx.RequestError` - Connection failures, timeouts
- `httpx.HTTPStatusError` - HTTP 4xx/5xx responses
- `json.JSONDecodeError` - Response parsing errors

Let unexpected exceptions propagate so they're visible in debug mode and Sentry.

## Affected Files

Commands that make HTTP requests and use broad exception handling:
- `src/magpie/cli/__init__.py` - version command with --server-version flag
- Other CLI commands in `src/magpie/cli/commands/` that interact with the API

## References

Original comment: https://github.com/SouthwestCCDC/magpie/pull/415#discussion_r2744280873

(AI-generated via Claude Code w/ Sonnet 4.5)

### PR DESCRIPTION
**Pull request by duplico**
## Summary

- Enhanced `with_network_error_handling` decorator to catch specific exception types
- Added `httpx.HTTPStatusError` handling for HTTP error status codes
- Added `json.JSONDecodeError` handling with user-friendly error messages
- Updated comprehensive decorator documentation
- Added test coverage for new exception types

## Context

Follow-up from PR #415 review where broad `except Exception` blocks were identified as problematic because they catch both expected errors (connection failures, timeouts, HTTP errors) and unexpected programming errors (attribute errors, bugs in parsing), making debugging harder and potentially hiding bugs.

### Exception Handling Improvements

1. **HTTPStatusError handling**: Added catch block for `httpx.HTTPStatusError` which is raised when using `raise_for_status()`. Routes through existing `handle_response_error()` for consistent formatting.

2. **JSONDecodeError handling**: Added catch block for `json.JSONDecodeError` to handle cases where server returns invalid JSON (e.g., error pages, misconfiguration). Provides helpful error message with hint about checking server logs.

3. **Documentation**: Updated decorator docstring to clearly document all exception types handled and note that unexpected exceptions are allowed to propagate for visibility in debug mode and Sentry.

### Design Rationale

- **Specific exception types**: Only catches known, expected exception types from httpx and json modules
- **Unexpected exceptions propagate**: Programming errors (AttributeError, KeyError, TypeError, etc.) are not caught, allowing them to be visible in debug mode and captured by Sentry for diagnosis
- **Consistent error formatting**: Uses existing error handling functions for consistent user experience
- **Comprehensive coverage**: Handles all expected error scenarios from HTTP requests and response parsing

## Test Plan

- [x] All existing unit tests pass (880 tests)
- [x] Added test for JSONDecodeError handling in decorator
- [x] Added test for HTTPStatusError handling in decorator
- [x] Added CLI runner integration test for JSONDecodeError exit codes
- [x] Added CLI runner integration test for HTTPStatusError exit codes
- [x] Verified linting passes (ruff)
- [x] Manual verification that unexpected exceptions still propagate (test_passes_through_non_network_errors)

## References

- Original issue: #417
- PR review comment: https://github.com/SouthwestCCDC/magpie/pull/415#discussion_r2744280873

Closes #417

Generated with Claude Code (Opus 4.6)

### CODE REVIEW

**Review by duplico — COMMENTED** [review]
# PR #481 Review - Refactor CLI exception handling

## Summary
- **3 issues found** (1 minor bug, 2 suggestions)
- **Severity**: MINOR (no blocking issues)
- **Recommendation**: REQUEST_CHANGES for the misleading exit code on line 299

## Critical Issues (Blocking)
None.

## Important Issues (Should Fix)

### 1. Misleading exit code in unreachable code (Line 299)

**File**: `src/magpie/cli/errors.py:299`

**Issue**: The HTTPStatusError handler has `raise SystemExit(ExitCode.NETWORK_ERROR)` which is:
1. Unreachable (handle_response_error never returns)
2. Wrong exit code (should be status-specific, not hardcoded to NETWORK_ERROR)

**Evidence**:
- The test `test_http_status_error_exit_code_with_runner` expects exit code 3 (NOT_FOUND) for 404
- The test PASSES, proving line 299 is never executed
- `handle_response_error` uses `http_status_to_exit_code()` to map status codes correctly

**Fix**: Either:
- Remove the line entirely since it's unreachable
- Or fix it to match the comment's intent (but still mark as unreachable):
  ```python
  raise SystemExit(http_status_to_exit_code(exc.response.status_code))  # pragma: no cover
  ```

**Why this matters**: Code reviewers and future maintainers will see NETWORK_ERROR and think 404s exit with code 2, when they actually exit with code 3. Misleading code is a maintenance hazard.

## Suggestions (Non-blocking)

### 2. Dead code - HTTPStatusError handling has no trigger path

**Context**: The entire HTTPStatusError exception handler (lines 285-299) is defensive code for a future scenario that doesn't currently exist.

**Evidence**:
- No CLI code uses `response.raise_for_status()` (checked entire src/magpie/cli/ and src/magpie/ctl/)
- All code manually checks `response.status_code` and calls `handle_response_error()`
- The only `raise_for_status()` call is in e2e test fixtures

**Not a bug, but worth noting**:
- This is defensive programming for future-proofing
- If someone adds `raise_for_status()` later, it will be handled correctly
- But currently increases test coverage requirements for untriggered code
- Consider adding a comment explaining this is future-proofing

### 3. Test coverage for JSON output mode

**Missing test**: No test for HTTPStatusError or JSONDecodeError in JSON output mode (--format=json).

**Current coverage**:
- Human mode: ✅ test_http_status_error_exit_code_with_runner
- JSON mode: ❌ No equivalent test

**Why it matters**: The decorator has different code paths for JSON vs human mode. While the underlying functions are tested, the decorator's integration with JSON mode for these new exception types is not.

**Suggested test**:
```python
def test_http_status_error_json_mode(self):
    """HTTPStatusError in JSON mode produces correct error output."""
    from unittest.mock import patch
    
    @click.command()
    @with_network_error_handling
    def test_command():
        mock_request = MagicMock()
        mock_response = MagicMock()
        mock_response.status_code = 404
        mock_response.json.return_value = {"detail": "Not found"}
        raise httpx.HTTPStatusError(
            "404 Not Found", request=mock_request, response=mock_response
        )
    
    runner = CliRunner()
    with patch('magpie.cli.errors.is_json_output', return_value=True):
        result = runner.invoke(test_command, ['--format=json'])
    
    assert result.exit_code == ExitCode.NOT_FOUND
    # Verify JSON error output structure
```

## Positive Observations

1. **Comprehensive test coverage**: Added 4 new tests covering both unit and integration levels
2. **Exception hierarchy verified**: HTTPStatusError and RequestError are siblings (not parent/child), so catch order doesn't matter
3. **Consistent with existing patterns**: Uses same error formatting functions as existing code
4. **Good documentation**: Docstring clearly lists all handled exception types
5. **Preserves debugging**: Unexpected exceptions still propagate to Sentry
6. **Exit codes correct**: Despite the misleading unreachable code, actual exit codes are correct (verified by tests)

## Test Verification

All 33 tests pass:
```
============================= test session starts ==============================
tests/unit/test_cli_network_errors.py::TestWithNetworkErrorHandlingDecorator::test_catches_json_decode_error PASSED [ 63%]
tests/unit/test_cli_network_errors.py::TestWithNetworkErrorHandlingDecorator::test_catches_http_status_error PASSED [ 66%]
tests/unit/test_cli_network_errors.py::TestCliRunnerIntegration::test_json_decode_error_exit_code_with_runner PASSED [ 96%]
tests/unit/test_cli_network_errors.py::TestCliRunnerIntegration::test_http_status_error_exit_code_with_runner PASSED [100%]
============================== 33 passed in 0.21s ==============================
```

## Review Checklist

- ✅ Clear, readable code with meaningful names
- ✅ Appropriate error handling (specific exceptions)
- ✅ Follows existing patterns in the codebase
- ✅ Type hints present on public functions
- ✅ Tests cover new functionality
- ⚠️ Edge cases handled (JSON mode not tested for new exceptions)
- ✅ Not over-mocked (tests real behavior)
- ✅ AI disclosure present

## Recommendation

**REQUEST_CHANGES** - Fix the misleading exit code on line 299 before merging.

The code is functionally correct (tests prove it), but the unreachable SystemExit has the wrong exit code, which will mislead future maintainers. This is a simple fix that prevents future confusion.

Suggestions #2 and #3 are nice-to-have improvements but not blocking.

---

(AI-generated review via Claude Code w/ Sonnet 4.5)

File: src/magpie/cli/errors.py [hunk 1]
```diff
@@ -272,10 +278,25 @@ def my_command(ctx):
...
+            token = None
+            if ctx and ctx.obj:
+                if hasattr(ctx.obj, "server"):
+                    server = ctx.obj.server
+                if hasattr(ctx.obj, "token"):
+                    token = ctx.obj.token
+
+            # Use handle_response_error for consistent error formatting
+            handle_response_error(exc.response, operation=func.__name__, token=token)
+            raise SystemExit(ExitCode.NETWORK_ERROR)  # pragma: no cover
```
Comment (duplico) [inline src/magpie/cli/errors.py]: **Issue: Misleading exit code in unreachable code**

This line has the wrong exit code - it should be `http_status_to_exit_code(exc.response.status_code)` not `ExitCode.NETWORK_ERROR`.

While this line is never executed (handle_response_error never returns), it's misleading to future maintainers. A 404 HTTPStatusError should conceptually exit with code 3 (NOT_FOUND), not code 2 (NETWORK_ERROR).

**Evidence**: The test `test_http_status_error_exit_code_with_runner` expects exit code 3 for a 404 and passes, proving this line is unreachable and the actual exit code comes from handle_response_error.

**Fix**: Either remove this line entirely or use:
```python
raise SystemExit(http_status_to_exit_code(exc.response.status_code))  # pragma: no cover
```

(AI-generated via Claude Code w/ Sonnet 4.5)

Comment (duplico) [reply src/magpie/cli/errors.py]: Fixed in 3dfcc4c8a6c0c96e59d76b60f48de64ac5b48dd8. Changed the unreachable line to use `http_status_to_exit_code(exc.response.status_code)` instead of the misleading hardcoded `NETWORK_ERROR` exit code. This ensures the code correctly indicates status-specific exit codes even though the line is never executed.

(AI-generated via Claude Code w/ Opus 4.6)

File: src/magpie/cli/errors.py [hunk 2]
```diff
@@ -272,10 +278,25 @@ def my_command(ctx):
...
+            # HTTP error responses (4xx, 5xx) raised by raise_for_status()
+            # Extract server and token from context if available
+            ctx = click.get_current_context(silent=True)
+            server = None
+            token = None
+            if ctx and ctx.obj:
+                if hasattr(ctx.obj, "server"):
+                    server = ctx.obj.server
+                if hasattr(ctx.obj, "token"):
+                    token = ctx.obj.token
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/cli/errors.py]: Fixed in 3dfcc4c8a6c0c96e59d76b60f48de64ac5b48dd8. Removed the unused `server` variable extraction as suggested. The HTTPStatusError handler now only extracts the token from context.

(AI-generated via Claude Code w/ Opus 4.6)

File: src/magpie/cli/errors.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/cli/errors.py]: Fixed in 3dfcc4c8a6c0c96e59d76b60f48de64ac5b48dd8. Changed to use `http_status_to_exit_code(exc.response.status_code)` which correctly maps HTTP status codes to appropriate exit codes (404 → NOT_FOUND, 401/403 → AUTH_ERROR, etc.). Also added a comment explaining this handler is defensive/future-proofing.

(AI-generated via Claude Code w/ Opus 4.6)

**Comment by duplico** [comment]
## Review Feedback Addressed

All review findings have been addressed in commit 3dfcc4c8a6c0c96e59d76b60f48de64ac5b48dd8:

### IMPORTANT Issue (Fixed)
1. **Misleading exit code on line 299**: Changed from hardcoded `NETWORK_ERROR` to `http_status_to_exit_code(exc.response.status_code)` to correctly indicate status-specific exit codes

### SUGGESTION Issues (Addressed)
2. **Dead code comment**: Added explanatory comment that the HTTPStatusError handler is defensive/future-proofing (currently no CLI code calls `raise_for_status()`)
3. **Missing JSON mode tests**: Added two new tests:
   - `test_json_decode_error_json_mode`: Verifies JSONDecodeError in JSON mode produces correct error output
   - `test_http_status_error_json_mode`: Verifies HTTPStatusError in JSON mode produces correct error output

### Additional Cleanup
- Removed unused `server` variable in HTTPStatusError handler (per Copilot review comment)

All tests pass (35/35, up from 33). Linting and formatting checks pass.

(AI-generated via Claude Code w/ Opus 4.6)
